linalg/densemat.cpp

```diff
@@ -3500,4 +3502,88 @@ DenseTensor &DenseTensor::operator=(double c)
    return *this;
 }
 
+void BatchLUFactor(DenseTensor &Mlu, Array<int> &P, const double TOL)
+{
+   const int m = Mlu.SizeI();
+   const int NE = Mlu.SizeK();
+   P.SetSize(m*NE);
+
+   auto data_all = mfem::Reshape(Mlu.ReadWrite(), m, m, NE);
+   auto ipiv_all = mfem::Reshape(P.Write(), m, NE);
+   Array<bool> pivot_flag(1);
+   pivot_flag[0] = true;
+   bool *d_pivot_flag = pivot_flag.ReadWrite();
+
+   MFEM_FORALL(e, NE,
+   {
+      for (int i = 0; i < m; i++)
+      {
+         // pivoting
+         {
+            int piv = i;
+            double a = fabs(data_all(piv,i,e));
+            for (int j = i+1; j < m; j++)
+            {
+               const double b = fabs(data_all(j,i,e));
+               if (b > a)
+               {
+                  a = b;
+                  piv = j;
+               }
+            }
+            ipiv_all(i,e) = piv;
+            if (piv != i)
+            {
+               // swap rows i and piv in both L and U parts
+               for (int j = 0; j < m; j++)
+               {
+                  mfem::kernels::internal::Swap<double>(data_all(i,j,e), data_all(piv,j,e));
+               }
+            }
+         }//pivot end
+
+         if (abs(data_all(i,i,e)) <= TOL)
+         {
+            d_pivot_flag[0] = false;
+         }
+
+         const double a_ii_inv = 1.0 / data_all(i,i,e);
+         for (int j = i+1; j < m; j++)
+         {
+            data_all(j,i,e) *= a_ii_inv;
+         }
+
+         for (int k = i+1; k < m; k++)
+         {
+            const double a_ik = data_all(i,k,e);
+            for (int j = i+1; j < m; j++)
+            {
+               data_all(j,k,e) -= a_ik * data_all(j,i,e);
+            }
+         }
+
+      }//m loop
+
+   });
+
+   MFEM_ASSERT(pivot_flag.HostRead()[0], "Batch LU factorization failed \n");
+}
+
+void BatchLUSolve(const DenseTensor &Mlu, const Array<int> &P, Vector &X)
+{
+
+   const int m = Mlu.SizeI();
+   const int NE = Mlu.SizeK();
+
+   auto data_all = mfem::Reshape(Mlu.Read(), m, m, NE);
+   auto piv_all = mfem::Reshape(P.Read(), m, NE);
+   auto x_all = mfem::Reshape(X.ReadWrite(), m, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      kernels::LUSolve(&data_all(0, 0,e), m, &piv_all(0, e), &x_all(0,e));
```

Comment:
You went back to the pointer interface?...

Reply:
You can make the signature take `(const int e, DeviceTensor<3> &data, DeviceTensor<2> &piv, DeviceTensor<2> &x)`

Reply:
I would think it might be better to put `LUSolve` in this file, any opinion @v-dobrev ?

Reply:
I added a comment above regarding this. I wanted to get some feedback on breaking away from the general pointer format used in the rest of the functions used in kernels.hpp. for example other functions in that file look like this: 
``` 
void MultABt(const int Aheight, const int Awidth, const int Bheight,                                                                                                                
             const TA *Adata, const TB *Bdata, TC *ABtdata)   
``` 
If introducing dense tensors in kernels.hpp is preferred we do could do that too.

Reply:
I'd just like to say that I do prefer the pointer interface here for the LU kernel. It allows us to use a wider range of data structures than if was changed to DenseTensor. I think that's an important feature to have for kernels such as this that can be used for a wide range of different applications.

Reply:
You can always create a `DeviceTensor` from a raw pointer and dimensions using the `Reshape` function.